### PR TITLE
[WIP] Add scrolling_frame

### PR DESCRIPTION
**Pull request by farnoy**
**Work in progress**

This is supposed to address #986, but I'm having a lot of trouble layouting it - separators are causing undesirable skips instead of scrolling steadily. It is constant-width and does not shift its neighbors.

It works with composite layouts, I did not do anything specific for Pango and I provided two modes of operation, infinity scroll, where it scrolls forever by one character, and normal scroll, where upon reaching the right edge, it snaps back to the beginning

### CODE REVIEW

**Comment by tobes** [comment]
@farnoy 

This doesn't run under python 2 due to `**item` in a dict, py3status supports python 2 and 3.  You also have some syntax issues see the travis test results for more details.

With the separators they might be difficult as i3bar does not want to show them after a block with no content so you might want to add a space to stand in for them at that point.  Another option is just to kill the separators (not sure how that would look)

> It is constant-width 

it does not seem to be at the end of the string maybe right pad it with spaces.

**Comment by farnoy** [comment]
@tobes 
I will adjust it for both python versions, sure.

Yeah the separators are sort of difficult, even if you don't want them, i3bar will put a white space in between your modules. It does not make a difference and I need to pad it with spaces either way.

Is there a way to unit-test or at least pass in other modules as children using module_test? I find it hard to test outside of actually using it and seeing the results scroll by.

**Comment by tobes** [comment]
> Is there a way to unit-test or at least pass in other modules as children using module_test? I find it hard to test outside of actually using it and seeing the results scroll by.

currently this isn't possible but it is something I'd like to add in the future

the best way for now is probably add lots of `self.py3.log(...)` calls to the module so you can see stuff in the logs

> i3bar will put a white space in between your modules.

adding 'separator': False should fix this if I understand your issue correctly

**Comment by lasers** [comment]
Hi @farnoy. Can I get your stance on this pull request? Will you still work on this? If not, is it okay to close them?

I played with this pull request and I think this behaves better with `cache_timeout = -1` too... to scroll a bit more nicer / faster.

I notice that I can have slightly hard time to parse things as I would need to internally recognize which module it is first before I can decide what to do with it. Reading left to right, I often have to acknowledge `PMS`, `MS` or `S` instead of just simple `DPMS`.

Am I reading or waiting for a module? I forget that some of them can be clicked too as I will view them as simple scrolling messages and/or information... and not something meant for me to interact with.

>It would be nice to have a module like frame, but giving it an upper bound on width (in # of characters) and having it scroll the text so it never takes more space than defined. Would this addition make sense to others, or is it just me? Maybe I'll try implementing it.

The better behavior may be to use some combination of `group` and `frame`.
First, configure `frame` first to use `{button}` placeholder as this is important.... Then put the non-essential modules in `frame` and the essential modules in `group` outside of `frame` to get that acceptable small-width desire. Also, if you plan on hiding the modules via `\?not_zero Mail {mail}` or such, they should sit outside too... especially if the modules are using `urgent` too. Anyhow, it's just one suggestion.

One more thing... If we got an `urgent` with this, it's often not clear immediately.
We would get `7| DPMS | XSCR | M` instead of `XSCR | DPMS | Mail 17`.

Overall, I think this feature looks good on paper but perhaps it's not a practical feature for the users to use... Let us know what you want to do with this and #986. Thanks. :-)

**Comment by farnoy** [comment]
Hey @lasers, I'm not working on this anymore and was unable to finish it because of edge cases around whitespaces, separators etc. I got more screens now, so I don't need to pack things together anymore.

**Comment by lasers** [comment]
@farnoy Ha. That's always a good solution too. Ten-four on the edge cases. Also, another reason why this is hard to parse... If you have `scrolling_frame` running with other modules, it will be hard to tell where it is... as it will constantly shuffle the separator bars too so it might be bad idea to use this with other modules. Thanks for letting us know.

@tobes I guess we can close this issue and pull request? Or maybe @farnoy do it.

**Comment by farnoy** [comment]
I'll let you decide, I am fine with closing it - it's just an unfinished prototype

**Comment by ultrabug** [comment]
Ok closing then, thanks again @farnoy
